crates/slot-arithmetic/src/lib.rs

```diff
@@ -215,51 +236,55 @@ impl EraHistory {
 
     pub fn slot_to_absolute_time(
         &self,
-        slot: u64,
+        slot: Slot,
         system_start: u64,
     ) -> Result<u64, TimeHorizonError> {
         self.slot_to_relative_time(slot).map(|t| system_start + t)
     }
 
-    pub fn relative_time_to_slot(&self, time: u64) -> Result<u64, TimeHorizonError> {
+    pub fn relative_time_to_slot(&self, time: u64) -> Result<Slot, TimeHorizonError> {
         for era in &self.eras {
             if era.start.time_ms > time {
                 return Err(TimeHorizonError::InvalidEraHistory);
             }
             if era.end.time_ms >= time {
                 let time_elapsed = time - era.start.time_ms;
                 let slots_elapsed = time_elapsed / era.params.slot_length;
-                let slot = era.start.slot + slots_elapsed;
+                let slot = era.start.slot
+                    + Slot {
+                        slot: slots_elapsed,
```

Comment:
This looks fishy: `slots_elapsed` is not a slot number, it is a distance in slot space. This is (and implies) the issues that drive the distinction between `Duration` and `Instant`, where two instants cannot be added, only subtracted (yielding a duration, of course).

I’d solve this by adding `Slot::offset_by(self, offset: u64) -> Self`, which should also validate that nothing wrong happens. Perhaps the offset should be restricted to u32, even.